### PR TITLE
[Proof of concept] Re-enable mod store and use a forum parser based DB

### PR DESCRIPTION
**Pull request by rubenwardy**
https://github.com/rubenwardy/mtmods4android_server

The server gets a list from Krock's mod search, and then validates it and improves it. This includes:
- Checking modname, authorname, download url.
- Checks that the download doesn't 404, but goes to a zip file.
- Fetches description.txt and depends.txt from Github based repos
- Download link points to the latest known commit. The commit hash is given and should be used to verify contents (if possible).

Not ready to be used due to:
- Needs to get a hash of the download to check for malicious modification
- Non-github downloads are still listed, but will be insecure due to no way to point to a commit
- When updating the commit the download link points to, the download should be virus scanned
- No category support yet
- No rating support yet
- No way for a mod owner to log into the db and correct things.
- MMDB uses modname as a primary key, where as my DB uses modname and author as a composite primary key. The ingame mod store needs to be changed to reflect this - my database supports multiple mods with the same name (eg: tenplus1/mobs vs pilzadam/mobs), whereas mmdb didn't

This is a proof of concept pr that shows that it is possible for a forum based parser to work well.

TL;DR: It all works, except it lacks features and security.

### CODE REVIEW

**Comment by sofar** [comment]
- can we extend mod.conf to serve as metadata storage?
- License info should be displayed at all times
- why is it bad that the downloads are from github? Just have the client perform a checksum check and let github serve as a cdn

**Comment by rubenwardy** [comment]
* I'm planning to read mod.conf in order to get tag info. It could be used to allow an author to correct some issues, but not all.
* It wasn't in the old one, afaik. It's reasonable hard to get that. Ideally it would be in mod.conf, but due to legacy you'd have to search in / look for license.txt, readme.md, copying, etc, and use a library to detect the license.
* This could be a good idea. You could make it point to a particular commit, and then let the server check the repo for new commits every so often to update. The client would then check the checksum on download.

**Comment by benrob0329** [comment]
Perhaps it could use libgit?

**Comment by sofar** [comment]
I wouldn't install mods from git. Just have the client download the master zip archive, instead.

**Comment by rubenwardy** [comment]
I agree, using git probably overengineering it.

**Comment by rubenwardy** [comment]
The mod database now points at the latest known commit, rather than master.

    github.com/rubenwardy/awards/archive/9cd25143f315694204c6bed68ebc331383dec535.zip

The sha hash of the commit is also stored, but I'm unsure if you could use that to verify the download - I may need to download the zip file and hash that download as well, for verification.

**Comment by t4im** [comment]
> The sha hash of the commit is also stored, but I'm unsure if you could use that to verify the download

The commit-sha is over the commit (a text file with the message and other shas). It should work even with missing history, like with shallow clones, but not with working directory snapshots that discard the commit itself.

> The mod database now points at the latest known commit, rather than master.

Would [github releases](https://developer.github.com/v3/repos/releases/) be an option, if set?
Sadly neither way includes git submodules yet. :-(

**Comment by rubenwardy** [comment]
I'm planning on letting the mod.conf in the default branch specify a download tag. That could be a branch name, a commit hash or a tag.

Ie, the mod.conf file on the default (master) branch:

   default_git_download_ref = stable-1.2

**Comment by iqualfragile** [comment]
@rubenwardy the primary key should not cause any problems. if you serialize in a compatible format the / in the name should not be a problem, the only thing that might happen would be display "problems" as in the title is shown as pilzadam/mobs.

**Comment by nerzhul** [comment]
@rubenwardy is this poc useful or can we close it ?
